File: 0115-distinct-subsequences/0115-distinct-subsequences.py
class Solution:
    def numDistinct(self, s: str, t: str) -> int:
        self.s = s
        self.t = t
        self.n = len(s)
        self.m = len(t)
        
        self.dp = [[-1] * (self.m + 1) for _ in range(self.n + 1)]
        
        return self.solve(0, 0)
        
    def solve(self, i, j):
        if j == self.m:
            return 1
        if i == self.n:
            return 0
        
        if self.dp[i][j] != -1:
            return self.dp[i][j]
        
        if self.s[i] == self.t[j]:
            self.dp[i][j] = self.solve(i + 1, j + 1) + self.solve(i + 1, j)
        else:
            self.dp[i][j] = self.solve(i + 1, j)
        
        return self.dp[i][j]

File: 0115-distinct-subsequences/0115-distinct-subsequences.cpp
class Solution {
public:
    int dp[1001][1001];
    int numDistinct(string s, string t) {
        memset(dp,-1,sizeof(dp));
        return solve(s,t,0,0);
    }
    int solve(string &s,string &t,int i,int j){
        if(j==t.size()){
            return 1;
        }
        if(i==s.size()){
            return 0;
        }
        if(dp[i][j]!=-1){
            return dp[i][j];
        }
        if(s[i]==t[j]){
            dp[i][j]=solve(s,t,i+1,j+1)+solve(s,t,i+1,j);
        }
        else{
            dp[i][j]=solve(s,t,i+1,j);
        }
        return dp[i][j];
    }
};